certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,560 @@
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be executed again.
+rule scheduledExecutionCanBeExecutedOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    execute@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// User cannot become an executor for already scheduled execution, meaning that
+// a user has to be supplied as one of the executors when scheduling an execution,
+// to be able to execute this execution.
+rule cannotBecomeExecutorForAlreadyScheduledExecution(env e, method f) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    address user;
+    require(length < max_uint256);
+    require(id < length);
+    bool isExecutor = isExecutor(id, user);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    assert !isExecutor => !isExecutor(id, user),
+        "User became an executor after the scheduled execution has been scheduled.";
+}
+
+
+// STATUS - verified
+// An execution can be executed (`canExecute`) only when it has not yet been cancelled
+// or executed and its `executableAt` is not greater than current timestamp.
+rule whatCanBeExecuted(env e) {
+    uint256 id;
+    uint256 length = getSchedExeLength();
+    require(id < length);
+
+    bool canExecute = canExecute(e, id);
+
+    assert (
+        !getSchedExeExecuted(id) &&
+        !getSchedExeCancelled(id) &&
+        getSchedExeExecutableAt(id) < e.block.timestamp
+    ) => canExecute;
+}
+
+
+// STATUS - verified
+// A protected scheduled execution can be executed only by an executor.
+rule whoCanExecute(env e) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    require(id < length);
+
+    bool canExecute = canExecute(e, id);
+    bool isExecutor = isExecutor(id, e.msg.sender);
+    bool isProtected = getSchedExeProtected(id);
+    bool executedBefore = getSchedExeExecuted(id);
+
+    execute(e, id);
+
+    bool executedAfter = getSchedExeExecuted(id);
+
+    assert !executedBefore && executedAfter => isExecutor || !isProtected;
+    assert canExecute && isExecutor => executedAfter;
+}
+
+
+// STATUS - verified
+// to cancel user must be canceler and it is impossible to execute after canceled.
+rule cancelerCanCancel(env e) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    require(id < length);
+    bool isCanceler = isCanceler(id, e.msg.sender);
+
+    bool canceledBefore = getSchedExeCancelled(id);
+    bool executed = getSchedExeExecuted(id);
+
+    cancel(e, id);
+
+    bool canceledAfter = getSchedExeCancelled(id);
+
+    assert canceledAfter;
+    assert !canceledBefore => isCanceler;
+
+    execute@withrevert(e, id);
+    bool reverted = lastReverted;
+    assert canceledAfter => reverted;
```

Comment:
I thought it would be a good practice: When using lastReverted, always store it into reverted variable right after the function call so that you will never make the mistake of rewriting lastReverted before reasoning about it. But I agree, that in this case, it is more readable not to introduce the reverted variable.

Reply:
I think it's a safer approach to do as you did here @tadeas-kucera, otherwise you will forget to add it someday and it will cause, for example, false positive result. 
I presently found this issue during Lido review.